### PR TITLE
Add CardValidator to payment view model

### PR DESCRIPTION
**Pull request by ehab-al-cko**
[Jira ticket](https://checkout.atlassian.net/jira/software/c/projects/PIMOB/boards/988?modal=detail&selectedIssue=PIMOB-1191)
## Proposed changes

ExpiryDateView(environment: viewModel.environment) is public and needs to pass in the environment for merchants to add this to their UI. 

Instead the ExpiryDateView should take in a CardValidator in the init method. 

## Types of changes

What types of changes does your code introduce to Frames Ios?
_Put an `x` in the boxes that apply_

* [ ] Bugfix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [ ] Lint and unit tests pass locally with my changes
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by deepesh-vasthimal-cko — APPROVED** [review]

**Review by alex-ioja-yang-cko — CHANGES_REQUESTED** [review]
> ExpiryDateView(environment: viewModel.environment) is public and needs to pass in the environment for merchants to add this to their UI.
> Instead the ExpiryDateView should take in a CardValidator in the init method.


As the description states, looks like the card validator is replacing the environment. That property is now unused.

File: Source/UI/NewUI/PaymentForm/Factory/PaymentFormFactory.swift [hunk 1]
```diff
@@ -1,27 +1,29 @@
 import UIKit
+import Checkout
 
 public struct PaymentFormFactory {
+  public static func getPaymentFormViewController(environment: Environment = .live,
```
Comment (alex-ioja-yang-cko) [inline Source/UI/NewUI/PaymentForm/Factory/PaymentFormFactory.swift:5]: Could we unit test this method to check the factory creates a ViewController with all the expected properties?

File: Tests/UI/New UI/PaymentFormFactoryTests.swift [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+@testable import Frames
+
+class PaymentFormFactoryTests: XCTestCase {
+
+  func testGetPaymentFormViewController() {
+    let viewController = PaymentFormFactory.getPaymentFormViewController(
+      billingFormData: nil,
+      paymentFormStyle: nil,
+      billingFormStyle: nil)
+    XCTAssertNotNil(viewController)
```
Comment (alex-ioja-yang-cko) [inline Tests/UI/New UI/PaymentFormFactoryTests.swift:18]: Could we also dig into the properties of the controller to check that the visible properties are as expected?

Maybe also add tests for those properties not being nil and seeing the factory generates the VC with all the properties you would expect for the inputs?

Comment (alex-ioja-yang-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:18]: Had a look and the ViewModel inside VC is internal for reading, so you could check things like:
- viewController.viewModel. cardValidator
- viewController.viewModel. billingFormData
- viewController.viewModel. paymentForStyle
- viewController.viewModel. billingFormStyle

File: iOS Example Frame/iOS Example Frame/MainViewController.swift [hunk 3]
```diff
@@ -66,10 +69,10 @@ final class MainViewController: UIViewController, CardViewControllerDelegate, Th
                           country: Self.countryGB)
         let name = "User Custom 1"
         let billingForm = BillingForm(name: name, address: address, phone: phone)
-        let paymentFormViewController = PaymentFormFactory.getPaymentFormViewController(
-            billingFormData: billingForm,
-            paymentFormStyle: Style.Custom1.paymentForm,
-            billingFormStyle: Style.Custom1.billingForm)
+        let paymentFormViewController = PaymentFormFactory.getPaymentFormViewController( environment: .live,
```
Comment (deepesh-vasthimal-cko) [inline iOS Example Frame/iOS Example Frame/MainViewController.swift:72]: Change it to sandbox.

File: iOS Example Frame/iOS Example Frame/MainViewController.swift [hunk 4]
```diff
@@ -85,10 +88,10 @@ final class MainViewController: UIViewController, CardViewControllerDelegate, Th
         let name = "User Custom 2"
         let billingForm = BillingForm(name: name, address: address, phone: nil)
 
-        let paymentFormViewController = PaymentFormFactory.getPaymentFormViewController(
-            billingFormData: billingForm,
-            paymentFormStyle: Style.Custom2.paymentForm,
-            billingFormStyle: Style.Custom2.billingForm)
+        let paymentFormViewController = PaymentFormFactory.getPaymentFormViewController( environment: .live,
```
Comment (deepesh-vasthimal-cko) [inline iOS Example Frame/iOS Example Frame/MainViewController.swift:91]: Change it to sandbox.

File: Tests/UI/New UI/PaymentFormFactoryTests.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    let name = "User 1"
+
+    let billingForm = BillingForm(name: name, address: address, phone: phone)
+
+    let viewController = PaymentFormFactory.getPaymentFormViewController(
+      billingFormData: billingForm,
+      paymentFormStyle: paymentFormStyle,
+      billingFormStyle: billingFormStyle)
+
+    XCTAssertNotNil(viewController)
```
Comment (alex-ioja-yang-cko) [inline Tests/UI/New UI/PaymentFormFactoryTests.swift:36]: I see that `PaymentFormFactory.getPaymentFormViewController` returns `UIViewController`, not `UIViewController?`.

I think its worth testing things that are possible, not Swift type safety approach. Swift language team probably covers that on their side.

File: Tests/UI/New UI/PaymentFormFactoryTests.swift [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    let paymentViewController = try XCTUnwrap(viewController as? PaymentViewController)
+
+    XCTAssertNotNil(paymentViewController.viewModel)
+    XCTAssertNotNil(paymentViewController.viewModel.billingFormData)
+    XCTAssertNotNil(paymentViewController.viewModel.billingFormStyle)
+    XCTAssertNotNil(paymentViewController.viewModel.paymentFormStyle)
```
[the remark this answers is not in the document]
Comment (ehab-al-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:43]: It will take a lot of time to make all objects are equatable. There are other tests to make sure all inside are tested.

[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:43]: > It will take a lot of time to make all objects are equatable.
Sometimes instead of making object equatable we can just compare the properties in separate assertions. It depends how many properties there are


> There are other tests to make sure all inside are tested.
There's no unit test to confirm the data provided to the factory is the data ending up inside the returned VC or its VM. There's also no unit test to confirm that any custom changes introduced in the factory are unit tested to prevent unintended changes. There may be tests for VC and/or the VM, but we're looking at the factory, its behaviour should also be validated and constrained by appropriate unit tests imo.

[the remark this answers is not in the document]
Comment (ehab-al-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:43]: This will take alot of time to check every value in all features in frames sdk. it will cost alot of time.

[the remark this answers is not in the document]
Comment (ehab-al-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:43]: @deepesh-vasthimal-cko what do you think?

**Review by alex-ioja-yang-cko — APPROVED** [review]

File: Tests/UI/New UI/PaymentFormFactoryTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Tests/UI/New UI/PaymentFormFactoryTests.swift:45]: Tried this myself, yes its not possible. Pretty disappointed as unit testing should be the goal of most design patterns and in this case this is a huge black hole we cannot enforce.
